### ISSUE TITLE
using more images for the same face, higher res texture

### ISSUE THREAD

**Issue by 2blackbar** [abdallahdib/NextFace#2]
hey i tested it, its ok but not much better than other 3d avatar apps out there that can also take sideviews.
Is it planned to use additional sideviews for mesh generation? from what i see ,the same male mesh is used on all models and the mesh itself is just slightly adapted, not that much, so in the end its almost like slapping photo on base mesh without artifacts. sideview is the same with all people.
It also takes awhile on 1080ti , would be nice if it could retain the colours of original image and produce material texture of 1024, let it compute mesh on256 but texture should be higher res
also theres an issue with landmarks, when face is wider than base mesh then landmarks places the dots on cheek, not on the border of the face... maybe cause its not perfect front view

**Comment by abdallahdib** [comment]
U can use NextFace on multiple images for same subject. just put ur images in the same folder and run optimizer.py with the additional flag: --sharedIdentity (images must have the same resolution, please refer to readme file for more details). this should improve the estimated geometry and reflectance than using a single image. 

for processing time, this is first order gradient descent method and yes it is slow (I have a deep based approach that runs at real time which i will be releasing soon... please refer to limitations in the readme file)

> _would be nice if it could retain the colours of original image and produce material texture of 1024,_
the method disentangles skin attributes (diffuse, specular, roughness) when combined together with rendering equation u should retreive the original image colors ( i suggest to read more about rendering, materials, BRDF, rendering..)
nevertheless, if you want to acheive ur desired effect, u can remove lighting and specular/roughness maps  from the 
 optimization and overfits the diffuse map (remove all regularizers for  diffuse)  

NextFace relies on landmarks  from https://github.com/1adrianb/face-alignment which are unfortunately not accurate and this may degrades the reconstruction quality.

**Comment by abdallahdib** [comment]
Not sure if what deepfacelab is using. face-alignment still the most used one and deepfacelab may already be using it.

**Comment by xqyd** [comment]
> U can use NextFace on multiple images for same subject. just put ur images in the same folder and run optimizer.py with the additional flag: --sharedIdentity (images must have the same resolution, please refer to readme file for more details). this should improve the estimated geometry and reflectance than using a single image.
> 
> for processing time, this is first order gradient descent method and yes it is slow (I have a deep based approach that runs at real time which i will be releasing soon... please refer to limitations in the readme file)
> 
> > _would be nice if it could retain the colours of original image and produce material texture of 1024,_
> 
> the method disentangles skin attributes (diffuse, specular, roughness) when combined together with rendering equation u should retreive the original image colors ( i suggest to read more about rendering, materials, BRDF, rendering..) nevertheless, if you want to acheive ur desired effect, u can remove lighting and specular/roughness maps from the optimization and overfits the diffuse map (remove all regularizers for diffuse)
> 
> > also theres an issue with landmarks, when face is wider than base mesh then landmarks places the dots on cheek
> 
> NextFace relies on landmarks from https://github.com/1adrianb/face-alignment which are unfortunately not accurate and this may degrades the reconstruction quality.

Thanks for your great work! Upon your suggestion, I've used --sharedIdentity flag for the same person's front view and side view photo. However, I still get two generated faces instead of one. Since these two photos belong to the same person, could it possible to fuse these two meshes into a single face, with which the estimated shape is close to the front-view identity and the side-view identity at the same time.

**Comment by abdallahdib** [comment]
Hi, when using --sharedIdentity it do exactly what u want and u will get a single obj file and a single diffuse/specular/roughness for all images.

**Comment by xqyd** [comment]
Hi, I am confused. I've add `print("diffuseAlbedo.shape[0] = ", diffuseAlbedo.shape[0])` to double check it. And the output is 2, which means 
`        for i in range(diffuseAlbedo.shape[0]):`
`              saveObj(outputDir + prefix + '/mesh' + str(i) + '.obj',`
has been called twice to output mesh0.obj and mesh1.obj respectively.

And here's my console output:
**python optimizer.py --sharedIdentity --input ./input_batch --output ./output_batch**
loading optim config from:  ./optimConfig.ini
Loading Basel Face Model 2017 from ./baselMorphableModel/morphableModel-2017.pickle...
loading mesh normals...
loading uv parametrization...
loading landmarks association file...
creating sampler...
loading images from path:  ./input_batch
loading image from path:  ./input_batch/frame216.jpg
[INFO] resizing input image to fit: 256 px resolution...
loading image from path:  ./input_batch/frame30.jpg
[INFO] resizing input image to fit: 256 px resolution...
detecting landmarks...
init camera pose...
1/3 => Optimizing head pose and expressions using landmarks...
100%|█████████████████████████████████████████████████████████████████████████████| 2000/2000 [00:06<00:00, 286.11it/s]
2/3 => Optimizing shape, statistical albedos, expression, head pose and scene light...
100%|████████████████████████████████████████████████████████████████████████████████| 401/401 [03:45<00:00,  1.78it/s]
3/3 => finetuning albedos, shape, expression, head pose and scene light...
100%|████████████████████████████████████████████████████████████████████████████████| 101/101 [00:59<00:00,  1.71it/s]
took 4.86 minutes to optimize
saving to: ' ./output_batch/input_batch/ '. hold on...
**diffuseAlbedo.shape[0] =  2**

**Comment by abdallahdib** [comment]
so yes, i think it is my fault it was not clear.

when using multiple images for the same person, the optimizer will use the same shape identity and albedo across all images, however the optimizer will produce an obj for each view (image) because the expression can change from image to another. but all the exported obj files share the same shape identity. that could be tricky to grasp i know. but to think about it, the geometry is composed of two components: 
- neutral mesh ( face without any expression). this is shared across all images.
- expression ( smile, sad ...) . the expression can change from an image to another

so to summarize, when u use --sharedIdentity, the neutral mesh is shared across all images but the expression is different for each image. which makes sense in fact, because our face exist in neutral pose and can present different expressions...

now for diffuseAlbedo.shape[0] = 2, this is true but u can notice the diffuseAlbedo is the same for all images. u can  also notice that the library produce a single uv maps  for all images

**Comment by xqyd** [comment]
![image](https://user-images.githubusercontent.com/16912042/164721323-3917f087-7d61-41be-a028-6d436ef69c33.png)
As shown in this Figure, the left one is generated by the side view image, and the right one is generated by the front view image. What I am trying to get is that a person with low bridge of nose, should have this consistency whenever it is viewed from various angles. So what I really want is a result that fuses both front-view-generated-face and side-view-generated-face.

**Comment by abdallahdib** [comment]
u are not looking to the mesh from the same point of view. I will add an export ofthe neutal face (without expression) optimized across all the images. hope this will help

**Comment by ZhenyanSun** [comment]
> it was not clear.

hello abdallahdib，
I want to know how can I get an export of the neutal face optimized of all the images.
